spec.bs

```diff
@@ -1844,11 +1867,26 @@ Expand the section of the navigation/traversal response handling which deals wit
 <h3 id="navigate-event-traversal-patches">History traversal updates</h3>
 
 <div algorithm="apply the history step">
-  Modify the [=apply the history step=] algorithm as follows. Near the end of the algorithm, inside the queued task over each global (currently step 14.10), before potentially unloading or activating the history entry, add the following step:
+  Modify the [=apply the history step=] algorithm as follows. Near the start of the algorithm, before checking if unloading is user-canceled (currently step 5), add the following steps:
+  1. Let <i>targetEntry</i> be the result of [=getting the target history entry=] given <i>traversable</i> and <i>targetStep.</i>
+  1. If <i>targetEntry</i> does not equal <i>traversable</i>'s [=navigable/current session history entry=], then:
+    1. Let <i>continue</i> be the result of [=Firing a traversal navigate event=] at <i>traversable</i>'s [=navigable/active window=]'s [=Window/navigation API=] with <i>[=fire a traversal navigate event/destinationEntry=]</i> set to <i>targetEntry</i> and <i>[=fire a traversal navigate event/userInvolvement=]</i> set to <var ignore>userInvolvement</var>
```

Comment:
I've switched to this model, but I'm not sure it's entirely right. The cure draft implementation intentionally fires the navigate event for the traversable navigable first, so that the navigate event doesn't fire anywhere else if the traversal has been aborted.

Reply:
Hmm OK, yes, I think I gave bad advice here. So how exactly does the task queuing and thread-hopping work in the implementation? I can think of a few variants, and it's probably easier to ask you than to decipher the code...

1.
  - Browser process queues a task to the main Window to fire the navigate event + beforeunload
  - If not canceled, the browser process queues tasks to all other Windows to fire navigate event + beforeunload
1.
  - Browser process queues a task to the main Window to fire the navigate event
  - If not canceled, the browser process queues tasks to all Windows to fire navigate event (with a guard to not fire it twice on the main window) + beforeunload

plus more possibilities...?

We need to reify something similar in the spec, basically.

Reply:
The implementation does the first option. The main Window fires navigate+beforeunload, and if allowed to proceed, queues tasks for all other Windows to run navigate+beforeunload as needed.

I've added a separate step for the main Window, lmk what you think of that approach.

Reply:
OK, getting there, but...

- The patches have gotten too excessive :(. Can you instead rewrite the whole "check if unloading is user-canceled" algorithm, perhaps using something like "(Insert current steps 6.1-6.8 here)" for the parts that do stay the same? And maybe rename it to "check if unloading is canceled" at the same time, since it's not just users now...

- Instead of passing in all the navigables, then looping over them all to find the one with null parent, you can just pass in _traversable_, which is exactly the right one.

- "get all navigables that might experience a cross-document traversal" does some important work, and I'm not 100% sure we can replace it with "get all navigables whose current session history entry will change or reload". Can you walk me through what you did there?

Reply:
Switched to a full rewrite.

Now takes "get all navigables that might experience a cross-document traversal" and _traversable_ as parameters. Still needs all the changing navigables internally, but it now does that by running "get all navigables whose current session history entry will change or reload" internally if _traversable_ is non-null, and otherwise just copying the passed-in "get all navigables that might experience a cross-document traversal".